### PR TITLE
Processing of ALL update types via events system

### PR DESCRIPTION
**Pull request by jonnywilliamson**

### CODE REVIEW

File: src/Laravel/config/telegram.php [hunk 1]
```diff
@@ -58,10 +58,20 @@
...
+
+
+                // (Optional). If you would like to process specific types of updates you may
+                // set the class to deal with each one here.
+                \Telegram\Bot\Events\MessageEvent::class   => [
+                ],
+                \Telegram\Bot\Events\PollEvent::class        => [
+                ],
+                \Telegram\Bot\Events\InlineQueryEvent::class => [
+                    \Telegram\Bot\Listeners\InlineQueryListener::class, // This is a demo. TO be Removed.
```
Comment (irazasyed) [inline src/Laravel/config/telegram.php:73]: Please remove demo code.

File: src/Listeners/InlineQueryListener.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+{
+
+    /**
+     * @param InlineQueryEvent $event
+     */
+    public function handle(InlineQueryEvent $event)
+    {
+        dd($event);
+    }
+}
```
Comment (irazasyed) [inline src/Listeners/InlineQueryListener.php:17]: Please remove the demo listener.

File: src/Listeners/ProcessUpdate.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+        try {
+            $event->getBot()->getEventFactory()->dispatch(
+                $eventName,
+                new $eventName($event->getBot(), $event->getUpdate())
+            );
+        } catch (Throwable $e) {
+            //TODO
+            // What should happen if there was a new update type added to telegram but
+            // we haven't updated the SDK yet?
```
Comment (irazasyed) [inline src/Listeners/ProcessUpdate.php:30]: Maybe we can dispatch a generic event here in that case. Let me test a few things. For now, let's leave this as is but just remove the comment.

File: src/Objects/Update.php [hunk 4]
```diff
@@ -82,47 +95,21 @@ public function detectType(): ?string
...
-            case 'poll_answer':
-                return $this->poll_answer;
-        }
+        return $this->{$this->detectType()};
+    }
 
-        return $this;
+    public function getEvent()
+    {
+        return 'Telegram\\Bot\\Events\\' . Str::ucfirst(Str::camel($this->detectType())) . "Event";
```
Comment (irazasyed) [inline src/Objects/Update.php:112]: Please change to `Str::studly()` method here.

Comment (irazasyed) [reply src/Objects/Update.php:112]: Just use this:

```php
'Telegram\Bot\Events\\' . Str::studly($this->detectType()) . 'Event'
```

Changed:
- Change to `studly` method.
- No need to escape slashes for every word.
- Single quote for `Event` at the end to make it consistent.

**Comment by jonnywilliamson** [comment]
I re-did all of this and then forced pushed this branch again.

I'm actually much much happier with it. It feels so much cleaner and lighter and I think easier to understand.

Hope you think the same!

File: src/Objects/Update.php (hunk 4 above)
Comment (jonnywilliamson) [reply src/Objects/Update.php]: Force push done. All this was changed.

File: src/Listeners/ProcessUpdate.php (hunk 3 above)
Comment (jonnywilliamson) [reply src/Listeners/ProcessUpdate.php]: Don't need to anymore! It's much simpler.

File: src/Laravel/config/telegram.php (hunk 1 above)
Comment (jonnywilliamson) [reply src/Laravel/config/telegram.php]: Done.

File: src/Listeners/InlineQueryListener.php (hunk 2 above)
Comment (jonnywilliamson) [reply src/Listeners/InlineQueryListener.php]: Done.

**Comment by jonnywilliamson** [comment]
Oh.

One massive item I have left out is WHERE to place the command processing!

I did this on purpose. It's likely to go back in for the `update.received` event. But it's worth a chat about this.

**Pull request merged**